runtime/compiler/optimizer/EscapeAnalysis.cpp

```diff
@@ -7371,117 +7365,167 @@ void TR_EscapeAnalysis::heapifyForColdBlocks(Candidate *candidate)
+               i += intIncrVal;
+//                }
+            }
+         }
+      else // if (candidate->_origKind == TR::newvalue)
+         {
+         TR_OpaqueClassBlock *valueClass = (TR_OpaqueClassBlock*)candidate->_class;
+         const TR::TypeLayout* typeLayout = comp()->typeLayout(valueClass);
+         size_t fieldCount = typeLayout->count();
+
+         TR_ASSERT_FATAL(candidate->_fields != NULL || fieldCount == 0, "candidate %p of kind newvalue is expected to have non-NULL candidate->_fields if it has a non-zero fieldCount %d\n", candidate->_node, fieldCount);
+
+         if (trace())
+            traceMsg(comp(), "Updating heap allocation node with fieldCount == %d\n", fieldCount);
+
+         TR::Node *newValueHeapificationNode = heapAllocation->getFirstChild();
+
+         for (size_t idx = 0; idx < fieldCount; idx++)
             {
+            const TR::TypeLayoutEntry &fieldEntry = typeLayout->entry(idx);
+
+            TR::Node *stackFieldLoad = NULL;
             int32_t j;
-            int32_t fieldSize = 0;
+
             for (j = candidate->_fields->size()-1; j >= 0; j--)
                {
                FieldInfo &field = candidate->_fields->element(j);
-               fieldSize = field._size;
-               if (field._offset == i &&
+
+               if (field._offset == fieldEntry._offset &&
                    field._symRef &&
-                   (candidate->isContiguousAllocation() || field._symRef->getSymbol()->isAuto()) &&
-                   (candidate->_origKind == TR::New || candidate->_origKind == TR::newvalue))
+                   (candidate->isContiguousAllocation() || field._symRef->getSymbol()->isAuto()))
                   {
                   TR::DataType type = field._symRef->getSymbol()->getDataType();
 
-                  TR::Node *stackFieldLoad = NULL;
                   if (!candidate->isContiguousAllocation())
-                     stackFieldLoad = TR::Node::createWithSymRef(heapAllocation, comp()->il.opCodeForDirectLoad(type), 0, field._symRef);
-                  else
-                     stackFieldLoad = TR::Node::createWithSymRef(comp()->il.opCodeForIndirectLoad(type), 1, 1, stackAllocation, field._symRef);
-
-                  TR::Node *heapFieldStore = NULL;
-                  TR::TreeTop *translateTT = NULL;
-                  if (stackFieldLoad->getDataType() == TR::Address)
                      {
-                     heapFieldStore = TR::Node::createWithSymRef(TR::awrtbari, 3, 3, heapAllocation->getFirstChild(), stackFieldLoad, heapAllocation->getFirstChild(), field.fieldSymRef());
-                     if (comp()->useCompressedPointers())
-                        {
-                        translateTT = TR::TreeTop::create(comp(), TR::Node::createCompressedRefsAnchor(heapFieldStore), NULL, NULL);
-                        }
+                     stackFieldLoad = TR::Node::createWithSymRef(newValueHeapificationNode, comp()->il.opCodeForDirectLoad(type), 0, field._symRef);
```

Comment:
Can this be reached for a flattened field ? If so, will having an auto of that type work ?

Reply:
All of the symrefs for flattened fields will be the ultimate "leaf" fields, so they will all be of primitive types or reference fields that are not flattened.